User claims added or updated (#21345)
dotnet/AspNetCore.Docs: aspnetcore/security/authentication/social/additional-claims.md

## Add and update user claims

Claims are copied from external providers to the user database on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app, call [SignInManager.RefreshSignInAsync](xref:Microsoft.AspNetCore.Identity.SignInManager%601) on a user to force the generation of a new authentication cookie.

In the Development environment working with test user accounts, you can simply delete and recreate the user account. For production systems, new claims added to the app can be backfilled into user accounts. After [scaffolding the `ExternalLogin` page](xref:security/authentication/scaffold-identity) into the app at `Areas/Pages/Identity/Account/Manage`, add the following code to the `ExternalLoginModel` in the `ExternalLogin.cshtml.cs` file.

Add a dictionary of added claims. Use the dictionary keys to hold the claim types, and use the values to hold a default value. Add the following line to the top of the class. The following example assumes that one claim is added for the user's Google picture with a generic headshot image as the default value:

```csharp

> private readonly IDictionary<string, string> _addedClaims = 

Reviewer: Generalizing from add to add-or-update.
```suggestion
private readonly IDictionary<string, string> _claimsToSync = 
```

    new Dictionary<string, string>()
    {
        { "urn:google:picture", "https://localhost:5001/headshot.png" },
    };
```

Replace the default code of the `OnGetCallbackAsync` method with the following code. The code loops through the claims dictionary. Claims are added (backfilled) or updated for each user. When claims are added or updated, the user sign-in is refreshed using the <xref:Microsoft.AspNetCore.Identity.SignInManager%601>, preserving the existing authentication properties (`AuthenticationProperties`).

```csharp
public async Task<IActionResult> OnGetCallbackAsync(
    string returnUrl = null, string remoteError = null)
{
    returnUrl = returnUrl ?? Url.Content("~/");
    if (remoteError != null)
    {
        ErrorMessage = $"Error from external provider: {remoteError}";
        return RedirectToPage("./Login", new {ReturnUrl = returnUrl });
    }

>     var info = await _signInManager.GetExternalLoginInfoAsync();

Reviewer: ```suggestion
    var externalLogin = await _signInManager.GetExternalLoginInfoAsync();
```

Author: @Tratcher  ... This is template driven, right? ...

https://github.com/dotnet/aspnetcore/blob/main/src/Identity/UI/src/Areas/Identity/Pages/V4/Account/ExternalLogin.cshtml.cs#L137

U sure that we should change it?

Reviewer 2: Yeah I'd match the scaffolded code over clarity

Reviewer: Darn. Maybe we should change the template...

Reviewer 2: Yeah that's fair, feel free to file an issue or PR for that

Author: Change it back to `externalLogin`?

If you know that's what you'll go with, I'd rather not come back to this given the workload 🏃😅.

Reviewer: Don't worry about it for now, any change wouldn't happen until 6.0.

Reviewer 2: Yeah this looks great to me,!

Author: I'll merge this when the build system comes back online.

    if (info == null)
    {
        ErrorMessage = "Error loading external login information.";
        return RedirectToPage("./Login", new { ReturnUrl = returnUrl });
    }

    // Sign in the user with this external login provider if the user already has a 
    // login.
    var result = await _signInManager.ExternalLoginSignInAsync(info.LoginProvider, 
        info.ProviderKey, isPersistent: false, bypassTwoFactor : true);
    if (result.Succeeded)
    {
        _logger.LogInformation("{Name} logged in with {LoginProvider} provider.", 
            info.Principal.Identity.Name, info.LoginProvider);

        if (_addedClaims.Count > 0)
        {
            var user = await _userManager.FindByLoginAsync(info.LoginProvider, 
                info.ProviderKey);
            var userClaims = await _userManager.GetClaimsAsync(user);
            bool refreshSignIn = false;

            foreach (var addedClaim in _addedClaims)
            {
                var userClaim = userClaims
                    .FirstOrDefault(c => c.Type == addedClaim.Key);

                if (info.Principal.HasClaim(c => c.Type == addedClaim.Key))
                {

>                     var principalClaim = info.Principal.FindFirst(addedClaim.Key);

Reviewer: ```suggestion
                    var externalClaim = info.Principal.FindFirst(addedClaim.Key);
```

                    if (userClaim == null)
                    {
                        await _userManager.AddClaimAsync(user, 
                            new Claim(addedClaim.Key, principalClaim.Value));
                        refreshSignIn = true;
                    }
                    else if (userClaim.Value != principalClaim.Value)
                    {
                        await _userManager
                            .ReplaceClaimAsync(user, userClaim, principalClaim);
                        refreshSignIn = true;
                    }
                }
                else if (userClaim == null)
                {

>                     await _userManager.AddClaimAsync(user, new Claim(addedClaim.Key, 

Reviewer: ```suggestion
                    // Fill with a default value
                    await _userManager.AddClaimAsync(user, new Claim(addedClaim.Key, 
```

                        addedClaim.Value));
                    refreshSignIn = true;
                }
            }

            if (refreshSignIn)
            {
                await _signInManager.RefreshSignInAsync(user);
            }
        }
                
        return LocalRedirect(returnUrl);
    }
    if (result.IsLockedOut)
    {
        return RedirectToPage("./Lockout");
    }
    else
    {
        // If the user does not have an account, then ask the user to create an 
        // account.
        ReturnUrl = returnUrl;
        ProviderDisplayName = info.ProviderDisplayName;
        if (info.Principal.HasClaim(c => c.Type == ClaimTypes.Email))
        {
            Input = new InputModel
            {
                Email = info.Principal.FindFirstValue(ClaimTypes.Email)
            };
        }
        return Page();
    }
}
```

> A similar approach is taken when claims change while a user is signed in but a backfill step isn't required. To update a user's claims, call the following on the user:

Author: But this line is meant to cover the **_not adding_** claim(s) scenario, where only replacing claim(s) and refreshing are required. I've angled this toward the dev refactoring the example to remove the `userClaim == null` pieces. I admit that I'm trying to punt 🏈 out of showing it. I think most devs can work it out easily enough from here. That's just my *guess* tho.

> A similar approach is taken when claims change while a user is signed in but a backfill step isn't required. To update a user's claims, call the following on the user:

Reviewer: *shrug* if the above code works and then they decide to remove part of it they're not using, that seems fine for them to sort out.

For more information, see <xref:Microsoft.AspNetCore.Authentication.OAuth.Claims>.

## Add claims to users

> Claims are added to the user on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app, call [SignInManager.RefreshSignInAsync](xref:Microsoft.AspNetCore.Identity.SignInManager%601) on a user to force the generation of a new authentication cookie.

Reviewer: ```suggestion
Claims are copied from external providers to the user database on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app, call [SignInManager.RefreshSignInAsync](xref:Microsoft.AspNetCore.Identity.SignInManager%601) on a user to force the generation of a new authentication cookie.
```

In the Development environment working with test user accounts, you can simply delete and recreate the user account.

## Update user claims

Claims that change after a user signs in aren't automatically updated. If claims change while a user is signed in, call:

* [UserManager.ReplaceClaimAsync](xref:Microsoft.AspNetCore.Identity.UserManager%601) on the user for claims stored in the identity database.
* [SignInManager.RefreshSignInAsync](xref:Microsoft.AspNetCore.Identity.SignInManager%601) on the user to force the generation of a new authentication cookie.

> 

Reviewer: This isn't giving much context on where these calls should be put. Do we have sample code to go with it?

Author: That's the question that I'm asking :point_up:. We don't cover these in the doc set. AFAIK, there's no sample for this scenario. 

It was discussed at https://github.com/dotnet/AspNetCore.Docs/pull/15491#discussion_r348664977, and that's where we left it. I'd need more guidance on how to address two scenarios that come to mind:

* The dev wants a utility method that they can run on all of the users to add/subtract claim(s) in the Identity dB.
* The devs wants `IUserClaimsPrincipalFactory` (?) to run automatically the next time a user signs in and updates them with the added claim(s).

... and that's if I'm even understanding the problem correctly. When we discussed this back in the day, I added claims ... set up the app ... but nothing was happening for my test user. You told me that the Identity dB needs to be updated (and currently signed in users). At the time, my quick resolution was merely to delete the test user and recreate them, which is fine for dev/test, not for production.

Author: > the app could on subsequent external logins call replace claim with the updated value of a claim. ... It shouldn't be necessary for them to have a custom principal factory for this.

But, that's not the same as adding a new claim from the provider for all users, right?

Author: I'm going to try again here now. Reminder on the scenario ...

1. App is created with user's stored locally.
2. Follow our guidance add an external social IP with some claims set up for social IP information about users.
3. Run the app ... register a user. Confirm that the user is able to auth and that the user has claims that were set up for the app.
4. The app is then modified ... the app is coded to set a **_new claim_** from Google about users (i.e., *Map user data keys and create claims* is followed to add a new one to the app).
5. Run the app again. Sign the test user in. Initially, the user **_doesn't get this new claim_** because "**_Claims are added to the user on first registration, not on signin._**"

That's where we left it. Thus far, your remarks are still a bit *ungrokable* 😵 to me, but I'll hack around a bit with your remarks in a test app to see if I can figure out *where* these calls go. Again, I think there are two potential solutions ...

* The dev runs a utility method that does ✨ *magic things* ✨ to all registered users in one shot to add the new claim. The next time any user signs in, *Bingo!* ... they have the new claim.
* The next time a user signs in, some ✨ *magic things* ✨ happen automatically for just that user to add the new claim. *Bingo!* ... they have the new claim.

Sorry if I've overly-simplified/overly-complicated the scenario. We never used social IPs at my last employer with .NET (pre-Core, of course), and claims-based Identity (IIRC) was only used for B2C-based apps during my final years there, which was *Identity Dreamy*&trade; :smile: to use. I recall my auth stress level dropped quite a bit using B2C.

Author: Rubber 🦆 working through this. Adding a few **_note to self_** cross-links.

This blog post covers what I'm facing ...

https://haacked.com/archive/2019/07/16/external-claims/

... and following the breadcrumbs ...

https://github.com/aspnet/Identity/issues/628

... leads to ...

https://github.com/dotnet/aspnetcore/issues/5771

... which was mentioned again on my issue for this PR ... and then trying to reach this ...

https://github.com/aspnet/AuthSamples/tree/dev/samples/Identity.ExternalClaims

... which is :boom: now but I think I can see here ...

https://github.com/aspnet/AuthSamples/commit/404105bd191f2e973d4befb668f1310d2fd82701

Author: Ok ..... rubber 🦆 says we're getting closer ...

```csharp
public async Task<IActionResult> OnGetCallbackAsync(string returnUrl = null, string remoteError = null)
{
    ...

    if (result.Succeeded)
    {
        _logger.LogInformation("{Name} logged in with {LoginProvider} provider.", info.Principal.Identity.Name, info.LoginProvider);

        if (info.Principal.HasClaim(c => c.Type == "urn:google:picture"))
        {
            _logger.LogInformation($"########### {info.Principal.FindFirst("urn:google:picture").Value}");

            var user = await _userManager.FindByLoginAsync(info.LoginProvider, info.ProviderKey);
            var claims = await _userManager.GetClaimsAsync(user);
            var claim = claims.FirstOrDefault(c => c.Type == "urn:google:picture");
            await _userManager.ReplaceClaimAsync(user, claim, info.Principal.FindFirst("urn:google:picture"));

            await _signInManager.RefreshSignInAsync(user);
        }

        return LocalRedirect(returnUrl);
    }
    ...
}
```

Author: @HaoK ... If I'm hack'in anything good here, **_A approach_** (not necessarily *THE approach* to show) is to ...

1. Run a migration on the dB and add a claim entry to `AspNetUserClaims` for every user. In my testing use case for adding the user's Google avatar image URL to the app, I can add the following to the dB for each user ...
   * UserId: The UserId (GUID key) for the user in the dB.
   * ClaimType: `urn:google:picture`
   * ClaimValue: `https://localhost:5001/foilhead.jpg` (a dummy value given that if this is left `NULL` in the dB it leads to a :boom: when a user signs into the app)

1. Drop an image for that into the app's `wwwroot` folder just in case a claim with a real Google image URL can't be assigned to the value of the claim.
 
   `wwwroot/foilhead.jpg`:

     ![foilhead](https://user-images.githubusercontent.com/1622880/108103148-2034f180-704f-11eb-8e6c-d58ec8711d33.jpg)

   *btw - Yeah, it was a hell of a party.* :beers:

1. Use that code in my last comment :point_up:.

   Of course, it also uses the current guidance for [map a key and make a claim](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/additional-claims?view=aspnetcore-5.0#map-user-data-keys-and-create-claims) that we have. In my test case ...

   ```csharp
   options.ClaimActions.MapJsonKey("urn:google:picture", "picture", "url");
   ```

   ... and the bits we show for `OnPostConfirmationAsync` ...

   ```csharp
   if (info.Principal.HasClaim(c => c.Type == "urn:google:picture"))
   {
       await _userManager.AddClaimAsync(user, info.Principal.FindFirst("urn:google:picture"));
   }
   ```

This approach seems to work. Is it sane :+1: ... or totally 😵 **_INSANE_** 😵?

Reviewer 2: Right, I would think its more reasonable to 'backfill' update the user's claim on login(option 2) over (periodically) synchronizing claims via a migration/import.

And the code you wrote looks reasonable, one minor thing is maybe you would also want to check if the claim is the same before replacing/refreshing the sign in.  So to summarize what that's doing, is on long, you check to see if any claims need to be updated, and do so then (and refresh the cookie if needed)  There's always this handshake to switch from the external login cookie to the real application cookie, so it doesn't seem unreasonable to look at the claims during that process.  @blowdart @Tratcher thoughts?

Author: I think I follow what you're suggesting. I think that I see the next step for the 'backfill' part and avoiding replace/refresh when there's no work to do for a given user that already has it and with a correct value. I'll work it further forward on Wednesday morning.

Reviewer: Refreshing/backfilling claims on login makes sense. Can that be done generically for a set of claims rather than for one claim at a time?

Author: Ok ... put an :eye: on this. This covers every scenario AFAICT in testing (but it might lack code economy and/or perf) ...

* Claim is present for the principal and ...

  * The user has no claim at all: Do a `AddClaimAsync` + `RefreshSignInAsync`.
  * The user has a claim and the values of the two match: Do a `ReplaceClaimAsync` + `RefreshSignInAsync`.
  * The user and principal claim values match: 🛑 **Drop out** ... just use what was loaded from the user claim.

* Claim not present in principal and ...

  * The user has a claim stored that they can use: 🛑 **Drop out** ... just use what was loaded from the user claim.
  * Both no principal claim and no existing user claim: Give them a fake headshot claim (`headshot.jpg`) ...

    ![headshot2](https://user-images.githubusercontent.com/1622880/108205472-bbc77000-70ea-11eb-8a43-8bdf8f2154d5.jpg)

    *I'll let the Baltimore Police Department arrest report and my attorney explain that one!* :smile:

**WRT what we'll eventually show**: If it's possible to generalize and cover a group of claims, that's cool. If so tho, it would be best to provide the full code to me because I need to get back to Blazor issues asap, where I have a *ton of work* ⛰️⛏️ to do. Otherwise, my plan is to show+describe an example of adding one specific claim, probably this one, and leaving it to the dev to take it further.

```csharp
var result = await _signInManager.ExternalLoginSignInAsync(info.LoginProvider, info.ProviderKey, isPersistent: false, bypassTwoFactor : true);
if (result.Succeeded)
{
    _logger.LogInformation("{Name} logged in with {LoginProvider} provider.", info.Principal.Identity.Name, info.LoginProvider);

    var user = await _userManager.FindByLoginAsync(info.LoginProvider, info.ProviderKey);
    var userClaims = await _userManager.GetClaimsAsync(user);
    var userPictureClaim = userClaims.FirstOrDefault(c => c.Type == "urn:google:picture");

    if (info.Principal.HasClaim(c => c.Type == "urn:google:picture"))
    {
        var principalPictureClaim = info.Principal.FindFirst("urn:google:picture");

        if (userPictureClaim == null)
        {
            await _userManager.AddClaimAsync(user, new Claim("urn:google:picture", principalPictureClaim.Value));
            await _signInManager.RefreshSignInAsync(user);
        }
        else if (userPictureClaim.Value != principalPictureClaim.Value)
        {
            await _userManager.ReplaceClaimAsync(user, userPictureClaim, principalPictureClaim);
            await _signInManager.RefreshSignInAsync(user);
        }
    }
    else if (userPictureClaim == null)
    {
        await _userManager.AddClaimAsync(user, new Claim("urn:google:picture", "https://localhost:5001/headshot.jpg"));
        await _signInManager.RefreshSignInAsync(user);
    }

    return LocalRedirect(returnUrl);
}
```
